### PR TITLE
run actions also with collier

### PR DESCRIPTION
**Pull request by wkotlarski**

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -90,7 +91,7 @@ jobs:
...
 
       - name: Configure
-        run: docker exec builder bash -c "cd FlexibleSUSY && if [ \"$LOOPLIBRARY\" = \"looptools\" ]; then export FL=\"--enable-looptools --with-looptools-libdir=/LoopTools-$LOOPTOOLS_VERSION/lib64 --with-looptools-incdir=/LoopTools-$LOOPTOOLS_VERSION/include\"; fi && ./configure --with-cxx=$CXXCOMPILER --with-math-cmd=/usr/local/Wolfram/WolframEngine/12.1/Executables/math --with-optional-modules=test --with-models=\$(echo $MODELS| tr ' ' ',') \$FL"
+        run: docker exec builder bash -c "cd FlexibleSUSY && if [ \"$LOOPLIBRARY\" = \"looptools\" ]; then export FL=\"--with-loop-libraries=looptools --with-looptools-libdir=/LoopTools-$LOOPTOOLS_VERSION/lib64 --with-looptools-incdir=/LoopTools-$LOOPTOOLS_VERSION/include\"; elif [ \"$LOOPLIBRARY\" = \"collier\" ]; then export FL=\"--with-loop-libraries=collier --with-collier-libdir=/COLLIER/lib --with-collier-incdir=/COLLIER/include\"; fi && ./configure --with-cxx=$CXXCOMPILER --with-math-cmd=/usr/local/Wolfram/WolframEngine/12.1/Executables/math --with-optional-modules=test --with-models=\$(echo $MODELS| tr ' ' ',') \$FL"
```
Comment (Expander) [inline .github/workflows/tests.yml:94]: Is it possible to write this long line over multiple lines? I'm not smart enough to read this. :)

**Comment by Expander** [comment]
Sorry for the dumb question, but what is the branch `feature-actions_with_library_interface` for?

**Comment by wkotlarski** [comment]
I want to adapt the test runner after Vova introduced loop library interface. This is related to issue #198.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (wkotlarski) [reply .github/workflows/tests.yml:94]: Maybe. I'm constantly thinking about it. The core problem is that we need to run the docker image explicitly. And then we run every command from the base image on docker image as `docker exec builder bash -c "..."`. A much nicer solution is telling in an action to directly run every command in the container as
```
runs-on: ubuntu-18.04
container: navir/opensuseleap-for-flexiblesusy
```
The problem is with mathematica. The Wolfram Engine then rejects my license. I don't know why and whether it's possible to fix it at all. I can reduce the length of the line a little even without it, though.

**Comment by wkotlarski** [comment]
So in all tests we should now call `Loop_library::set(-1)` and I should export the `FLEXIBLESUSY_LOOP_LIBRARY` variable? To save time, could I configure FS with LoopTools and Collier and run make once but tests few times (however many values do `FLEXIBLESUSY_LOOP_LIBRARY` take)?

**Comment by Expander** [comment]
Hmm, adding `Loop_library::set(-1)` to every test feels a bit of an unnecessary mess to me. @uukhas, what is your opinion of making `-1` the default?

I think in principle you could run
~~~
export FLEXIBLESUSY_LOOP_LIBRARY=0
make -j2 -k all-test
make clean-test-log
export FLEXIBLESUSY_LOOP_LIBRARY=1
make -j2 -k execute-compiled-tests execute-shell-tests execute-meta-tests
make clean-test-log
export FLEXIBLESUSY_LOOP_LIBRARY=2
make -j2 -k execute-compiled-tests execute-shell-tests execute-meta-tests
~~~
In this way the nothing would be re-compiled.

I'm not sure what happens in the mu to e conversion then.

Possible values for `FLEXIBLESUSY_LOOP_LIBRARY` can be: -1, 0, 1, 2, 3 (FFlite)

File: .github/workflows/tests.yml (hunk 1 above)
Comment (Expander) [reply .github/workflows/tests.yml:95]: I've played around a bit and found a site to convert YAML to JSON. It tells me that when I write
~~~
run:
  - >
    cmd "if [ -n "$file" ] ; then
    echo 'Hi there!' ;
    fi " | docker -
  - other_cmd
~~~
it gets translated to
~~~
{
   "run": [
      "cmd \"if [ -n \"$file\" ] ; then echo 'Hi there!' ; fi \" | docker -\n",
      "other_cmd"
   ]
}
~~~
So, I think it is possible to split the long line, even in the middle of the long string command that you're passing to docker.

**Comment by wkotlarski** [comment]
If `Loop_library::set(-1)` is not set, then env variable is not read, right? Am I missing something?

**Comment by Expander** [comment]
Yes, this is correct.

I've just checked: The default is "undefined" (which is not `-1`). I have the impression that in your case `-1` would be a sensible default. @uukhas is this correct?

**Comment by uukhas** [comment]
If the default is `-1`, then we are unable to use `.slha` files in order to set up a library.

**Comment by uukhas** [comment]
> So in all tests we should now call `Loop_library::set(-1)` and I should export the `FLEXIBLESUSY_LOOP_LIBRARY` variable? To save time, could I configure FS with LoopTools and Collier and run make once but tests few times (however many values do `FLEXIBLESUSY_LOOP_LIBRARY` take)? Also, what happens if the user configures the code without a dedicated loop library but then calls part of the code that requires it? Like mu to e conversion?

mu to e conversion can be calculated with all loop libraries. `softsusy`, for example, doesn't define some functions, so for them `loop_library` will return `nan`.

**Comment by Expander** [comment]
Sorry, I have a question/suggestion to discuss with you to help @wkotlarski :

What would happen if we'd make `-1` the default (not `0`)? We could come up with the following logic:

* If no value has been set, then the value is equal to the default `-1`. Then we proceed as follows:
  1. Look at the environment variable. If it has been set to `0`, ..., `3`, use the corresponding library.
  2. otherwise use `0`.
* If a value has been set that is not `-1`, use the corresponding loop library.

As far as I can see, this has the following advantages:

* We don't need to modify any tests and @wkotlarski can simply control the loop library through the environment variable
* We fall back to `0` (= SOFTSUSY) in any undefined case.

What do you think?

**Comment by uukhas** [comment]
The function called `::set` currently can be called only **once** in order to setup only one loop library per run. Using this **one** function user can define in which way he/she wants to setup the library.
 
`::set(-1)` says "look at environmental variable and in **any** case set it up immediately to something". No further looks for setups loaded from `.slha`. 

`::set(0|1|2|3|<any int>)` says "looking only at my argument setup the library immediately". 

By using `::set(-1)` we already provide some state to a loop library and directly after call of this function we know, that the value is set up via _environment_ variable only.

Can someone, please, explain to me, how a safe setup with **both simultaneously** env./slha can be implemented in this case?

**Comment by Expander** [comment]
As far as I can see, the `::set(int)` function is indeed called only once, namely in
~~~
@ModelName@_spectrum_generator_interface<T>::set_settings(
   const Spectrum_generator_settings& settings_)
~~~
where the `settings_` have been read from the input (SLHA or Mathematica).

But I don't see a reason why one could not (at least in principle) modify the `::set(int)` function in such a way that it can be called multiple times. I understand that you've written it in this way, but why can it not be changed? :) Maybe I'm missing something.

**Comment by uukhas** [comment]
this is exactly what I have. And this has to be called anyway somewhere, right?

**Comment by uukhas** [comment]
> I understand that you've written it in this way, but why can it not be changed? :) Maybe I'm missing something.

So, you want to setup a singleton multiple times?

**Comment by Expander** [comment]
Your function seems to be a bit different: Initially the type is `undefined`. In the `::set()` function you have:
~~~
if (type is undefined) {
   <define it>
}
~~~
I.e. in your case the loop library can only be set once, namely when the type is `undefined`. (Please correct me if I'm wrong.)

> So, you want to setup a singleton multiple times?

Yes. `Loop_library::lib_` etc. are not `const`, so they can in principle be changed. Or am I missing something? The singleton pattern does not imply const-ness, right?

Although, actually I'm not proposing to change the state later on. :) I think I'm proposing to have a different default value, namely `-1`. It could work like this:

1. FS starts. The initial `type_` is `undefined`.
2. The SLHA file is read -> `new_type` is read, which is either -1 (= if unset), 0, 1, 2 or 3. (This is different from before. Before we had -1, 0 (= if unset), 1, 2 or 3.)
3. The `::set(new_type)` is called with the `new_type` from 2. (Your `::set()` function is kept as it is.)

Now the point is: If the SLHA file does not contain `FlexibleSUSY[31]`, then `new_type == -1`. This again means that FS first looks in the environment variables (as you've written it). If the environment variable is set to a sensible value, that value is used. If it is set to a disallowed value, we fall back to `0`. In this approach the loop library is only set once.

This has the advantage that if we remove all `FlexibleSUSY[31]` entries from the SLHA files use in the tests, then @wkotlarski can control the loop library by the environment variable. In other words: We are not using SOFTSUSY all the time. If I understand correctly, this is what @wkotlarski  had in mind. Or am I missing something?

**Comment by uukhas** [comment]
I was a little bit confused. Actually, the main point is that we do not want to call `set(-1)` explicitly. But we want this behavior. I can just leave `set` as it is right now, but slightly change get in order to call `set(-1)`. Then one can just call any function and it on the first call will get the value from environment.

**Comment by uukhas** [comment]
I have no rights to make a change on this branch, so here is the diff:
```diff
diff --git a/src/loop_libraries/loop_library.cpp b/src/loop_libraries/loop_library.cpp
index bd635ab2a..e21c33632 100644
--- a/src/loop_libraries/loop_library.cpp
+++ b/src/loop_libraries/loop_library.cpp
@@ -119,7 +119,7 @@ looplibrary::Loop_library_interface& Loop_library::get()
    if (Loop_library::type_ == Loop_library::Library::Undefined) {
       ERROR("Loop library should be initialized before first usage.\n"
             "Setting default library.");
-      Loop_library::set_default();
+      Loop_library::set(-1);
    }
    return *Loop_library::lib_;
 }
```

**Comment by Expander** [comment]
Ok, now I think I see what you're thinking about. :) You also had the C++ tests in mind, which don't work on SLHA input, of course.

Does your proposed change work in the SLHA and C++ test case as discussed?

**Comment by wkotlarski** [comment]
Apart from tests, I believe that observables should check if the proper loop library is in use. Imagine that a user configures FS with LoopTools but uses a card where the loop library is set to softsusy or not set at all. Some observables then will print NaN. This is not helpful. Every observable should check at runtime if we're running with library that allows it to be computed, and if not print a meaningful message. 

I also suggest to change the order of libraries. Move fflite from 3 to 1. So that every value larger than 1 guaranties a proper looplibrary. Otherwise we'd have to encode check like
```
if (looplibrary == 1 || looplibrary == 2)
```
And if we ever get a third loop library we'd have to go again throw every of those files and change it. In my concept, we only do
```
if (looplibrary >= 2)
```

**Comment by uukhas** [comment]
@Expander, I changed, as you wrote, also the default value to `-1` in case of the absence in `slha` file.

**Comment by uukhas** [comment]
> Apart from tests, I believe that observables should check if the proper loop library is in use. Imagine that a user configures FS with LoopTools but uses a card where the loop library is set to softsusy or not set at all. Some observables then will print NaN. This is not helpful. Every observable should check at runtime if we're running with library that allows it to be computed, and if not print a meaningful message.
> 
> I also suggest to change the order of libraries. Move fflite from 3 to 1. So that every value larger than 1 guaranties a proper looplibrary. Otherwise we'd have to encode check like
> 
> ```
> if (looplibrary == 1 || looplibrary == 2)
> ```
> 
> And if we ever get a third loop library we'd have to go again throw every of those files and change it. In my concept, we only do
> 
> ```
> if (looplibrary >= 2)
> ```

I would say that this is quite a nice point, which, probably, should be discussed in other enhancement issue.

**Comment by wkotlarski** [comment]
Can someone remind me how to, separatelly:
- run mathematica tests
- compile c++ tests (but don't run any of them)
- run compiled tests

**Comment by Expander** [comment]
Run Mathematica tests:
~~~
make execute-meta-tests
~~~
Build compiled tests:
~~~
make alltest
~~~
Run compiled tests:
~~~
execute-compiled-tests
~~~
Run shell script tests:
~~~
make execute-shell-tests
~~~
Maybe we should put this into the wiki?

**Comment by wkotlarski** [comment]
Do shell script tests depend on loop library? Should I run them multiple times?

**Comment by Expander** [comment]
I think in general yes, because they call the spectrum generators.

**Comment by wkotlarski** [comment]
> I have no rights to make a change on this branch, so here is the diff:
> 
> ```diff
> diff --git a/src/loop_libraries/loop_library.cpp b/src/loop_libraries/loop_library.cpp
> index bd635ab2a..e21c33632 100644
> --- a/src/loop_libraries/loop_library.cpp
> +++ b/src/loop_libraries/loop_library.cpp
> @@ -119,7 +119,7 @@ looplibrary::Loop_library_interface& Loop_library::get()
>     if (Loop_library::type_ == Loop_library::Library::Undefined) {
>        ERROR("Loop library should be initialized before first usage.\n"
>              "Setting default library.");
> -      Loop_library::set_default();
> +      Loop_library::set(-1);
>     }
>     return *Loop_library::lib_;
>  }
> ```
> 
> Addition after the next message in conversation:
> 
> ```diff
> diff --git a/src/spectrum_generator_settings.cpp b/src/spectrum_generator_settings.cpp
> index e6cbab7a9..c9c555a39 100644
> --- a/src/spectrum_generator_settings.cpp
> +++ b/src/spectrum_generator_settings.cpp
> @@ -330,7 +330,7 @@ void Spectrum_generator_settings::reset()
>     values[higgs_3loop_correction_at2_as]    = 1.;
>     values[higgs_3loop_correction_at3]       = 1.;
>     values[higgs_4loop_correction_at_as3]    = 1.;
> -   values[loop_library]                     = 0.; // 0 = (Softsusy)
> +   values[loop_library]                     = -1.; // -1 = (set via environment FLEXIBLESUSY_LOOP_LIBRARY)
>  }
>  
>  Loop_corrections Spectrum_generator_settings::get_loop_corrections() const
> ```

@uukhas I applied both patches. Is there anything else or will the tests now set the library according to shell variable?

**Comment by wkotlarski** [comment]
@uukhas what happens if variable `FLEXIBLESUSY_LOOP_LIBRARY` is not set? I get
```
Warning: Check FlexibleSUSY[31]:
Currently configured values are 0 (=softsusy), 1 (=COLLIER), 2 (=LoopTools).
Setting default library.
```
but what is the default? The test seems to not fail, but I believe it should.

**Comment by uukhas** [comment]
> @uukhas what happens if variable `FLEXIBLESUSY_LOOP_LIBRARY` is not set? I get
> 
> ```
> Warning: Check FlexibleSUSY[31]:
> Currently configured values are 0 (=softsusy), 1 (=COLLIER), 2 (=LoopTools).
> Setting default library.
> ```
> 
> but what is the default? The test seems to not fail, but I believe it should.

Is the variable is not set I assume that it is treated as empty string. It is converted to `0`, i.e. `SOFTSUSY`.

**Comment by uukhas** [comment]
> @uukhas I applied both patches. Is there anything else or will the tests now set the library according to shell variable?

It seems that these changes are enough to make given above changes.

**Comment by wkotlarski** [comment]
> > @uukhas what happens if variable `FLEXIBLESUSY_LOOP_LIBRARY` is not set? I get
> > ```
> > Warning: Check FlexibleSUSY[31]:
> > Currently configured values are 0 (=softsusy), 1 (=COLLIER), 2 (=LoopTools).
> > Setting default library.
> > ```
> > 
> > 
> > but what is the default? The test seems to not fail, but I believe it should.
> 
> If the variable is not set I assume that it is treated as empty string. It is converted to `0`, i.e. `SOFTSUSY`.

I'm no sure if this is a good idea. It does hide the stuff from developer. @Expander, your idea?

**Comment by Expander** [comment]
Personally, I like the way it is done now, because from my point of view `FLEXIBLESUSY_LOOP_LIBRARY` is primarily for developers to switch between the libraries w/o modifying any code. With this mind setting, if `FLEXIBLESUSY_LOOP_LIBRARY` is undefined, I find it perfectly reasonable to fall back to the default case (= SOFTSUSY). This would correspond to the normal use(r) case, where `FLEXIBLESUSY_LOOP_LIBRARY` is unset and all spectrum generators are just run as normal.

Concerning the tests, you are right, that the tests are not aware of any change of `FLEXIBLESUSY_LOOP_LIBRARY`. I'd propose to run:
~~~
FLEXIBLESUSY_LOOP_LIBRARY=0
make execute-compiled-tests
make clean-test-log # removes all test output files (.xml and .log)
FLEXIBLESUSY_LOOP_LIBRARY=1
make execute-compiled-tests
make clean-test-log
~~~

**Comment by wkotlarski** [comment]
@Expander, thanks a lot! I'll implement it. 

Sorry for bothering you again, but I'm not very fluent in reading makefiles. Some includes are not passed correctly
```
Building test/test_pv_crosschecks.o
g++ -Iconfig -Isrc -Ifflite -Islhaea -Idoc -Imeta -Itemplates -Imodels/SM -Itest -I. -std=c++14 -O2 -fPIC -c test/test_pv_crosschecks.cpp -o test/test_pv_crosschecks.o
In file included from test/test_pv_crosschecks.cpp:29:
src/wrappers.hpp:31:10: fatal error: Eigen/Core: Nie ma takiego pliku ani katalogu
   31 | #include <Eigen/Core>
      |          ^~~~~~~~~~~~
```

**Comment by Expander** [comment]
Thanks for the hint! It should be fixed in commit 1d175db84

**Comment by wkotlarski** [comment]
> Personally, I like the way it is done now, because from my point of view `FLEXIBLESUSY_LOOP_LIBRARY` is primarily for developers to switch between the libraries w/o modifying any code. With this mind setting, if `FLEXIBLESUSY_LOOP_LIBRARY` is undefined, I find it perfectly reasonable to fall back to the default case (= SOFTSUSY). This would correspond to the normal use(r) case, where `FLEXIBLESUSY_LOOP_LIBRARY` is unset and all spectrum generators are just run as normal.
> 
> Concerning the tests, you are right, that the tests are not aware of any change of `FLEXIBLESUSY_LOOP_LIBRARY`. I'd propose to run:
> 
> ```
> FLEXIBLESUSY_LOOP_LIBRARY=0
> make execute-compiled-tests
> make clean-test-log # removes all test output files (.xml and .log)
> FLEXIBLESUSY_LOOP_LIBRARY=1
> make execute-compiled-tests
> make clean-test-log
> ```

Ok. But I'd propose to change the message
```
Warning: Check FlexibleSUSY[31]:
Currently configured values are 0 (=softsusy), 1 (=COLLIER), 2 (=LoopTools).
Setting default library.
```
so that it says what is the default that is actually set.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (wkotlarski) [reply .github/workflows/tests.yml]: I've split the line in b8d9a8032cc60fa57a12035feb52b5f995bf7405. Let me know if this is what you wanted.

**Comment by wkotlarski** [comment]
I'm now waiting for resolution of #217. After that we should be done with this branch.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (Expander) [reply .github/workflows/tests.yml]: Many thanks, this looks much better! :)

**Comment by wkotlarski** [comment]
I think we are done. If the tests pass, can I merge it?

**Comment by Expander** [comment]
From my point, feel free to merge it!

**Pull request merged**